### ISSUE TITLE
Months not normalized to years in Period

### ISSUE THREAD

**Issue by roji** [nodatime/nodatime#879]
While working with Period and normalization, I noticed that months aren't normalized into years: a Period of 13 months does not get normalized to 1 years 1 month, like how a Period of 70 minutes gets normalized to 1 hour 10 minutes.

I imagine that this is because of the Hebrew calendar's 13th leap month, but wanted to confirm this is so. If that is the case, would it makes sense to provide an overload of `Normalize()` that accepts a CalendarSystem?

**Comment by jskeet** [comment]
Yes, it's to avoid making too many assumptions about months in a year. Accepting a CalendarSystem wouldn't help, as you still wouldn't know what to do if you're given a Hebrew calendar system (you can't normalize it on the grounds of 13 months per year, as not all years have 13 months).

We could add an overload with an extra parameter as "number of months to assume per year"?

**Comment by roji** [comment]
I was thinking of accepting a CalendarSystem according to which you would normalize months to years or not - with non-Hebrew calendars it would be safe to perform that operation (99% of usage cases, I'd venture to say), with the Hebrew calendar you wouldn't.

I'm not so sure about adding a parameter with the number of months, it requires/assumes a lot of knowledge from the user about the intricacies of calendar system etc...

I'm not really stuck on this, it's more out of curiosity. Just to give some context, I'm working on Nodatime type support in the Npgsql driver for PostgreSQL - providing developers the option of having Npgsql read and write Nodatime types instead of the standard BCL types (https://github.com/npgsql/npgsql/issues/546). While transferring Period to and from PostgreSQL I ran into a small issue: the wire representation of PostgreSQL `interval` is months, days and microseconds. Years isn't a separate component, because I imagine they didn't take into account the Hebrew calendar case. In any case, Nodatime nicely normalizes the microseconds to milliseconds, seconds, minutes etc., but always returns Periods with year 0, which in my case doesn't seem very acceptable. For now I'm doing the normalization on my side, but it could be useful to communicate the fact that I want the 12-month per year assumption to Nodatime's normalizer, via a CalendarSystem etc.

**Comment by jskeet** [comment]
Hmm. I think it might be quite surprising for it silently not normalize. Let's leave this issue open to see what we can do for 2.1... I'm not sure there are any really nice options though.

**Comment by roji** [comment]
No problem, you really can close it if you don't feel like anything's missing.

**Comment by jskeet** [comment]
I think it's missing to some extent - I'm just not sure how to best expose it.

**Comment by abelbraaksma** [comment]
I'm hitting the same issue and added a `NormalizeAll` extension method. Durations happen to have an international standard and normalizing in that case simply means modulo 12 for months and div 12 for years. 

There are other calendars with 13 months but since in a period you can never know which years contain leap months I don't think this matters much. Alternatively people could add the period to a date with a certain calendar and retrieve the difference between new and old date. 

Since existing implementations may rely on the current behaviour, an overload that takes an enum (normalize months or not, or even with options for each property) seems a good and non-intrusive way forward. 

Note that the current `Normalize` is sub optimal, it raises overflow exceptions fairly quickly because it calls `TotalNanoseconds` internally which cannot contain the full allowed range. Simple modulo arithmetic would solve this bug. I can report that separately if you want.

**Comment by jskeet** [comment]
> Durations happen to have an international standard and normalizing in that case simply means modulo 12 for months and div 12 for years.

Durations/periods are calendar-neutral, and I really don't want to move away from that. Saying "I don't think this matters much" for other calendar systems isn't a good idea IMO. I don't want to present something as if it's universal and calendar-neutral when it's not.

I'll consider an overload accepting an enum, as well as other options. For example, another option would be to pass in a `LocalDate` to normalize it with respect to - that way we could normalize days as well as months. That feels cleaner than something which will normalize "13 months and 400 days" to "1 year, 1 month and 400 days" for example.

> Note that the current Normalize is sub optimal, it raises overflow exceptions fairly quickly because it calls TotalNanoseconds internally which cannot contain the full allowed range. Simple modulo arithmetic would solve this bug. I can report that separately if you want.

When you say "fairly quickly" - that should only happen if the time parts of the period add up to more than 292 years. (It doesn't include the date parts.) It's documented that this may happen, but I'd be surprised if you're actually running into this. Can you give an example of a situation where this has come up in real-world usage? I'm happy to change this if it's a genuine use case, but if this is just a matter of extreme testing, I'm less worried.

**Comment by glittle** [comment]
And the Badíʿ has 19 months in each year, though those months are regular and predictable.

**Comment by jskeet** [comment]
With no further progress on this issue in the last 1 1/2 years, I'm going to move the milestone to 3.1-consider.

On reflection, I suspect that a `Normalize(LocalDate origin)` method would be easy to write - along the lines of:

```csharp
public Period Normalize(LocalDate origin) =>
    Between(origin, origin + this);
```

That would normalize days, weeks, and months, all in a calendar-centric way. But I don't want to include it right now without further consideration.

**Comment by abelbraaksma** [comment]
> With no further progress on this issue in the last 1 1/2 years 

Apologies, I must've missed your previous comment,or somehow forgot about that. 

> Can you give an example of a situation where this has come up in real-world usage? 

My use case is that I use your library in an XPath compiler that has durations (daytime durations and year month durations) as native types. They are unbounded, but the spec allows a range that's implementation defined.

However, XPath and xslt is used a lot in processing historical documents. It's not uncommon that historical ranges exceed 292 years. I'm not suggesting this is a very common use case, but there's real world code out there that has durations that exceed the Common Era.

I can understand it has low priority, but perhaps you would except a PR that fixed this, provided it doesn't introduce backwards compatibility issues?

**Comment by jskeet** [comment]
@abelbraaksma: If you're happy with the proposal for `Normalize(LocalDate origin)` it's trivial to do, so no PR is required. Am I right in saying this is still non-urgent, so it would be reasonable to include in 3.1 rather than 3.0? If you're actually very keen to have this Real Soon Now, I could definitely include it in 3.0 by potentially taking a little longer to bring that out. (I was hoping to do 3.0 this week, but that's somewhat arbitrary.)

**Comment by abelbraaksma** [comment]
@jskeet, for the normalization I'm using a work around. I'd love to have a wider range than 292 days, but I'd understand if that requires more thought and perhaps more time (if you'd agree to it at all). 

I understand your current proposal addresses normalization based on a calendar and doesn't touch the overflow issues. Don't let 3.0 delay on my account, I'm happy to wait till 3.1 for this change.

**Comment by jskeet** [comment]
@abelbraaksma: Normalization of the form I suggested wouldn't go via `TotalNanoseconds`, so wouldn't have that restriction.

You can add an extension method right now:

```csharp
public static class PeriodExtensions
{
    // TODO: Add argument validation that period != null
    public Period Normalize(this Period period, LocalDate origin) =>
        Period.Between(origin, origin + period);
}
````

Indeed, if you're happy to adopt that for now, it would provide a great basis for feedback as to whether it meets all your needs.

**Comment by jskeet** [comment]
@abelbraaksma: Any more feedback on this? Given how simple the extension method is to write for those who do need it, I'm tempted to close at this point.

**Comment by abelbraaksma** [comment]
@jskeet, I agree it's a simple workaround, but it isn't simple to come to this conclusion, let alone think of this by yourself. If such method were to exist, it'd be much simpler to discover.

I think I used such workaround at the time, but I'll have to check, it's been a while ;).

**Comment by jskeet** [comment]
@abelbraaksma: So can I check that this at least meets your needs? There's no point in deciding whether or not to include it in a later release if it doesn't do what users want :)

**Comment by abelbraaksma** [comment]
@jskeet It does, but I'm wondering if there's a simpler way than using a dummy date. For instance, it may be more sensible to use a calendar or an enum?

Or have two overloads, one that automatically takes a local date of 'now' for the common scenario, and one that takes a user input, perhaps?

**Comment by jskeet** [comment]
No, I *definitely* want to discourage the use of "now" - I don't believe the result of normalizing "35 days" should depend on which month we're in right now, for example. It's far more than just the calendar system - the date itself is *very* important.

**Comment by abelbraaksma** [comment]
@jskeet, ok, good points, I understand. In that case, as written, I think your proposal is good as-is :).
